drivers/usb/usbstor_new/disk.c

```diff
@@ -650,6 +716,8 @@ USBSTOR_HandleDeviceControl(
 
         /* Hack for IoReadPartitionTable call in disk.sys */
         ScsiInquiryData->RemovableMedia = ((ScsiInquiryData->DeviceType != DIRECT_ACCESS_DEVICE) ? ((UFIInquiryResponse->RMB & 0x80) ? 1 : 0) : 0);
+        // should be:
+        //ScsiInquiryData->RemovableMedia = ((ScsiInquiryData->DeviceType == DIRECT_ACCESS_DEVICE) ? ((UFIInquiryResponse->RMB & 0x80) ? 1 : 0) : 0);
```

Comment:
This is quite inverted logic! What's actually happening in disk.sys that requires such an inverted logic?

Reply:
I would make it easier for this: ScsiInquiryData->RemovableMedia = 0; :)

In general, UFI is a "USB Floppy Interface" and therefore it is not entirely correct to use structures and names ("UFIInquiryResponse"), although they are similar to SCSI.

What is this hack done for?
The current storage device stack (in contrast to the new pnp) does not fill or process STORAGE_HOTPLUG_INFO. Btw, also in the PnP manager do not implement IoGetDeviceProperty (... DevicePropertyRemovalPolicy ...).

Reply:
Maybe @HeisSpiter would be interested to know about this possible disk.sys problem?

Reply:
Note: I've tested this on real hardware. When booting a USB key I have the DeviceType == 0 == DIRECT_ACCESS_DEVICE. Using the "should be" part gives me the correct value for RemovableMedia (== TRUE); while keeping the old code gives me the wrong value.
If I use the new code (or manually set RemovableMedia to TRUE) I then get a BSOD later due to some recursive code between HalpGetFullGeometry(), IOCTL_DISK_GET_PARTITION_INFO and UpdateRemovableGeometry() in disk.sys, and IoReadPartitionTable().